consensus/consensus_v2.go

```diff
@@ -679,13 +683,104 @@ func (consensus *Consensus) commitBlock(blk *types.Block, committedMsg *FBFTMess
 	return nil
 }
 
+func (consensus *Consensus) updateLeader() {
+	prev := consensus.GetLeaderPubKey()
+	epoch := consensus.GetCurEpoch()
+	curNumber := consensus.Blockchain.CurrentHeader().Number().Uint64()
+	if consensus.Blockchain.Config().IsLeaderRotation(epoch) {
+		leader := consensus.GetLeaderPubKey()
+		for i := uint64(0); i < 5; i++ {
+			header := consensus.Blockchain.GetHeaderByNumber(curNumber - i)
+			if header == nil {
+				return
+			}
+			// Previous epoch, we should not change leader.
+			if header.Epoch().Uint64() != epoch.Uint64() {
+				return
+			}
+			// Check if the same leader.
+			pub, err := chain.GetLeaderPubKeyFromCoinbase(consensus.Blockchain, header)
+			if err != nil {
+				utils.Logger().Error().Err(err).Msg("Failed to get leader public key from coinbase")
+				return
+			}
+			if !pub.Object.IsEqual(leader.Object) {
+				// Another leader.
+				return
+			}
+		}
+		// Passed all checks, we can change leader.
+		var (
+			wasFound bool
+			next     *bls.PublicKeyWrapper
+		)
+		// The same leader for N blocks.
+		if consensus.Blockchain.Config().IsAllowlistEpoch(epoch) {
+			wasFound, next = consensus.Decider.NthNextHmyExt(shard.Schedule.InstanceForEpoch(epoch), leader, 1)
+		} else {
+			wasFound, next = consensus.Decider.NthNextHmy(shard.Schedule.InstanceForEpoch(epoch), leader, 1)
+		}
+		if !wasFound {
+			utils.Logger().Error().Msg("Failed to get next leader")
+			return
+		} else {
+			consensus.SetLeaderPubKey(next)
+		}
+		if consensus.IsLeader() && !consensus.GetLeaderPubKey().Object.IsEqual(prev.Object) {
+			// leader changed
+			go func() {
+				consensus.ReadySignal <- SyncProposal
+			}()
+		}
+	}
+}
+
 // SetupForNewConsensus sets the state for new consensus
 func (consensus *Consensus) SetupForNewConsensus(blk *types.Block, committedMsg *FBFTMessage) {
 	atomic.StoreUint64(&consensus.blockNum, blk.NumberU64()+1)
 	consensus.SetCurBlockViewID(committedMsg.ViewID + 1)
+	if blk.IsLastBlockInEpoch() {
```

Comment:
why do this change? why not directly fetch the epoch from the proposed block? that should be already correct.